File: practice-projects/springboot4ddd/src/main/java/com/github/microwind/springboot4ddd/application/service/user/UserService.java
package com.github.microwind.springboot4ddd.application.service.user;

import com.github.microwind.springboot4ddd.application.command.user.CreateUserCommand;
import com.github.microwind.springboot4ddd.application.command.user.UpdateUserCommand;
import com.github.microwind.springboot4ddd.application.dto.user.UserDTO;
import com.github.microwind.springboot4ddd.application.port.CachePolicy;
import com.github.microwind.springboot4ddd.application.port.CacheService;
import com.github.microwind.springboot4ddd.domain.exception.EntityNotFoundException;
import com.github.microwind.springboot4ddd.domain.model.user.User;
import com.github.microwind.springboot4ddd.domain.page.PageRequest;
import com.github.microwind.springboot4ddd.domain.page.PageResult;
import com.github.microwind.springboot4ddd.domain.repository.user.UserRepository;
import com.github.microwind.springboot4ddd.domain.service.user.UserUniquenessChecker;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * 用户应用服务
 *
 * <p>仅负责用例编排与事务管理，业务规则下沉到聚合根 {@link User} 与领域服务
 * {@link UserUniquenessChecker}。输入用 application.command 包下的 Command，
 * 输出统一为 {@link UserDTO}，不引用 interfaces 层。
 *
 * @author jarry
 * @since 1.0.0
 */
@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(transactionManager = "userTransactionManager")
public class UserService {

    private final UserRepository userRepository;
    private final UserUniquenessChecker userUniquenessChecker;
    private final CacheService cacheService;

    public UserDTO createUser(CreateUserCommand command) {
        User user = User.register(
                userUniquenessChecker,
                command.getName(),
                command.getEmail(),
                command.getPhone(),
                command.getWechat(),
                command.getAddress()
        );
        User savedUser = userRepository.save(user);
        log.info("User created: id={}, name={}", savedUser.getId(), savedUser.getName());
        return toDTO(savedUser);
    }

    public UserDTO getUserById(Long id) {
        return cacheService.getOrSet(
                CachePolicy.USER_KEY_PREFIX + id,
                CachePolicy.USER_TTL,
                () -> {
                    User user = userRepository.findById(id)
                            .orElseThrow(() -> new EntityNotFoundException("User", "id", id));
                    return toDTO(user);
                }
        );
    }

    public UserDTO getUserByName(String name) {
        return cacheService.getOrSet(
                CachePolicy.USER_KEY_PREFIX + "name:" + name,
                CachePolicy.USER_TTL,
                () -> {
                    User user = userRepository.findByName(name)
                            .orElseThrow(() -> new EntityNotFoundException("User", "name", name));
                    return toDTO(user);
                }
        );
    }

    public List<UserDTO> getAllUsers() {
        return userRepository.findAll().stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public PageResult<UserDTO> getAllUsers(PageRequest pageRequest) {
        return userRepository.findAll(pageRequest).map(this::toDTO);
    }

    public UserDTO updateUser(Long id, UpdateUserCommand command) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("用户", "id", id));

        if (command.getEmail() != null) {
            user.changeEmail(userUniquenessChecker, command.getEmail());
        }
        if (command.getPhone() != null) {
            user.changePhone(command.getPhone());
        }
        if (command.getWechat() != null) {
            user.changeWechat(command.getWechat());
        }
        if (command.getAddress() != null) {
            user.changeAddress(command.getAddress());
        }

        User updatedUser = userRepository.update(user);
        log.info("User updated: id={}, name={}", updatedUser.getId(), updatedUser.getName());

        cacheService.delete(CachePolicy.USER_KEY_PREFIX + updatedUser.getId());
        return toDTO(updatedUser);
    }

    public void deleteUser(Long id) {
        if (!userRepository.findById(id).isPresent()) {
            throw new EntityNotFoundException("用户", "id", id);
        }
        userRepository.deleteById(id);
        log.info("User deleted: id={}", id);
        cacheService.delete(CachePolicy.USER_KEY_PREFIX + id);
    }

    private UserDTO toDTO(User user) {
        return UserDTO.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .phone(user.getPhone())
                .wechat(user.getWechat())
                .address(user.getAddress())
                .createdAt(user.getCreatedTime())
                .updatedAt(user.getUpdatedTime())
                .build();
    }
}

File: practice-projects/gin-ddd/internal/application/service/user/user_service.go
// Package user 用户应用服务。
package user

import (
	"context"

	userDTO "gin-ddd/internal/application/dto/user"
	domainErrors "gin-ddd/internal/domain/errors"
	"gin-ddd/internal/domain/event"
	userModel "gin-ddd/internal/domain/model/user"
	userDomain "gin-ddd/internal/domain/repository/user"
	userService "gin-ddd/internal/domain/service/user"
	"gin-ddd/pkg/utils"
)

// UserService 用户应用服务。
type UserService struct {
	userRepo       userDomain.UserRepository
	checker        userService.UserUniquenessChecker
	eventPublisher event.EventPublisher
	userTopic      string
}

// NewUserService 构造用户应用服务。
func NewUserService(
	userRepo userDomain.UserRepository,
	checker userService.UserUniquenessChecker,
	eventPublisher event.EventPublisher,
	userTopic string,
) *UserService {
	return &UserService{
		userRepo:       userRepo,
		checker:        checker,
		eventPublisher: eventPublisher,
		userTopic:      userTopic,
	}
}

// CreateUser 创建用户。唯一性校验下沉到聚合根 user.Register。
func (s *UserService) CreateUser(ctx context.Context, name, email, phone, address string) (*userDTO.UserDTO, error) {
	utils.GetLogger().Info("UserService.CreateUser 开始: name=%s, email=%s", name, email)

	newUser, err := userModel.Register(ctx, s.checker, name, email, phone, address)
	if err != nil {
		utils.GetLogger().Warn("用户注册失败: %v, name=%s, email=%s", err, name, email)
		return nil, err
	}

	if err := s.userRepo.Create(ctx, newUser); err != nil {
		utils.GetLogger().Error("用户持久化失败: %v, name=%s, email=%s", err, name, email)
		return nil, err
	}

	s.publishEvents(ctx, newUser.PullEvents())
	utils.GetLogger().Info("用户创建成功: id=%d, name=%s, email=%s", newUser.ID, name, email)
	return userDTO.ToDTO(newUser), nil
}

// GetUserByID 根据 ID 获取用户。
func (s *UserService) GetUserByID(ctx context.Context, id int64) (*userDTO.UserDTO, error) {
	u, err := s.userRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if u == nil {
		return nil, domainErrors.NewNotFound("用户", "id", id)
	}
	return userDTO.ToDTO(u), nil
}

// GetUserByName 根据用户名获取用户。
func (s *UserService) GetUserByName(ctx context.Context, name string) (*userDTO.UserDTO, error) {
	u, err := s.userRepo.FindByName(ctx, name)
	if err != nil {
		return nil, err
	}
	if u == nil {
		return nil, domainErrors.NewNotFound("用户", "name", name)
	}
	return userDTO.ToDTO(u), nil
}

// GetAllUsers 获取所有用户。
func (s *UserService) GetAllUsers(ctx context.Context) ([]*userDTO.UserDTO, error) {
	users, err := s.userRepo.FindAll(ctx)
	if err != nil {
		return nil, err
	}
	return userDTO.ToDTOs(users), nil
}

// UpdateEmail 更新用户邮箱。唯一性校验由 user.UpdateEmail 通过领域服务完成。
func (s *UserService) UpdateEmail(ctx context.Context, id int64, email string) error {
	u, err := s.userRepo.FindByID(ctx, id)
	if err != nil {
		return err
	}
	if u == nil {
		return domainErrors.NewNotFound("用户", "id", id)
	}
	if err := u.UpdateEmail(ctx, s.checker, email); err != nil {
		return err
	}
	if err := s.userRepo.Update(ctx, u); err != nil {
		return err
	}
	s.publishEvents(ctx, u.PullEvents())
	return nil
}

// UpdatePhone 更新用户手机号。
func (s *UserService) UpdatePhone(ctx context.Context, id int64, newPhone string) error {
	u, err := s.userRepo.FindByID(ctx, id)
	if err != nil {
		return err
	}
	if u == nil {
		return domainErrors.NewNotFound("用户", "id", id)
	}
	if err := u.UpdatePhone(newPhone); err != nil {
		return err
	}
	if err := s.userRepo.Update(ctx, u); err != nil {
		return err
	}
	s.publishEvents(ctx, u.PullEvents())
	return nil
}

// DeleteUser 删除用户。
func (s *UserService) DeleteUser(ctx context.Context, id int64) error {
	if err := s.userRepo.Delete(ctx, id); err != nil {
		utils.GetLogger().Error("用户删除失败: %v, id=%d", err, id)
		return err
	}
	utils.GetLogger().Info("用户删除成功: id=%d", id)
	return nil
}

func (s *UserService) publishEvents(ctx context.Context, events []event.DomainEvent) {
	for _, e := range events {
		if err := s.eventPublisher.Publish(ctx, s.userTopic, e); err != nil {
			utils.GetLogger().Error("[UserService] 发布领域事件失败: type=%s, err=%v", e.EventType(), err)
		}
	}
}
